calico.go

```diff
@@ -306,57 +306,68 @@ func cmdDel(args *skel.CmdArgs) error {
 		"Node":         nodename,
 	}).Info("Extracted identifiers")
 
-	// Always try to release the address. Don't deal with any errors till the endpoints are cleaned up.
-	fmt.Fprintf(os.Stderr, "Calico CNI releasing IP address\n")
-	logger.WithFields(log.Fields{"paths": os.Getenv("CNI_PATH"),
-		"type": conf.IPAM.Type}).Debug("Looking for IPAM plugin in paths")
+	calicoClient, err := CreateClient(conf)
+	if err != nil {
+		return err
+	}
 
-	// We need to replace "usePodCidr" with a valid, but dummy podCidr string with "host-local" IPAM.
-	if conf.IPAM.Type == "host-local" && strings.EqualFold(conf.IPAM.Subnet, "usePodCidr") {
+	ep := api.WorkloadEndpointMetadata{
+		Name:         args.IfName,
+		Node:         nodename,
+		Orchestrator: orchestrator,
+		Workload:     workload,
+	}
 
-		// host-local IPAM releases the IP by ContainerID, so podCidr isn't really used to release the IP.
-		// It just needs a valid CIDR, but it doesn't have to be the CIDR associated with the host.
-		dummyPodCidr := "0.0.0.0/0"
-		var stdinData map[string]interface{}
+	endpoint := &ep
 
-		if err := json.Unmarshal(args.StdinData, &stdinData); err != nil {
+	// Handle k8s specific bits of handling the DEL.
+	if orchestrator == "k8s" {
+		wep, err := k8s.CmdDelK8s(calicoClient, ep, args, logger)
+		if err != nil {
+			if err == k8s.ContainerIDMismatchErr {
+				// CNI_ContainerID does not match WorkloadEndpoint ActiveInstanceID so ignoring the DELETE cmd.
+				return nil
+			}
 			return err
 		}
 
-		logger.WithField("podCidr", dummyPodCidr).Info("Using a dummy podCidr to release the IP")
-		stdinData["ipam"].(map[string]interface{})["subnet"] = dummyPodCidr
+		if wep != nil {
+			endpoint = wep
+		}
+	}
 
-		args.StdinData, err = json.Marshal(stdinData)
-		if err != nil {
+	// Delete the WorkloadEndpoint object from the datastore.
+	if err = calicoClient.WorkloadEndpoints().Delete(*endpoint); err != nil {
```

Comment:
Also, although I can't think of an actual scenario where it is likely to happen, in the k8s code when you do have the resource update conflict I think you should retry as Tom suggests.  You need to retry from the Get() so that the revision and active instance ID are up to date.  If the WEP has had the active instance ID updated then you'll know about it and be able to log that with certainty.